clusterloader2/pkg/measurement/util/wait_for_pods.go

```diff
@@ -49,31 +54,69 @@ type WaitForPodOptions struct {
 // WaitForPods waits till desired number of pods is running.
 // Pods are be specified by namespace, field and/or label selectors.
 // If stopCh is closed before all pods are running, the error will be returned.
-func WaitForPods(clientSet clientset.Interface, stopCh <-chan struct{}, options *WaitForPodOptions) error {
-	ps, err := NewPodStore(clientSet, options.Selector)
-	if err != nil {
-		return fmt.Errorf("pod store creation error: %v", err)
+func WaitForPods(clientSet clientset.Interface, stopCh <-chan struct{}, options *WaitForPodOptions) (finalErr error) {
+	var lister func() []*corev1.Pod
+	if options.Selector.FieldSelector == "" && options.Lister != nil {
+		// This is more efficient, which is crucial when WaitForPods gets called
+		// for a single pod and a lot of those get created.
+		labelSelector, err := labels.Parse(options.Selector.LabelSelector)
+		if err != nil {
+			return fmt.Errorf("invalid WaitForPodOptions.Selector.LabelSelector: %v", err)
+		}
+		lister = func() []*corev1.Pod {
+			pods, err := options.Lister.Pods(options.Selector.Namespace).List(labelSelector)
+			if err != nil {
+				// Should never happen when Lister is a local cache.
+				panic(err)
+			}
+			return pods
+		}
+	} else {
+		// We have to create our own watch because filtering by field
+		// is implemented inside the apiserver or we don't have
+		// a shared PodLister.
+		ps, err := NewPodStore(clientSet, options.Selector)
+		if err != nil {
+			return fmt.Errorf("pod store creation error: %v", err)
+		}
+		defer ps.Stop()
+		lister = ps.List
 	}
-	defer ps.Stop()
 
-	oldPods := ps.List()
+	oldPods := lister()
 	scaling := uninitialized
 	var podsStatus PodsStartupStatus
 	var lastIsPodUpdatedError error
 
+	desiredPodCount := options.DesiredPodCount()
+	klog.V(2).Infof("%s: %s: start waiting for pod count %d", options.CallerName, options.Selector.String(), desiredPodCount)
+	defer func() {
+		if finalErr != nil {
+			klog.Errorf("%s: %s: failed: %v", options.CallerName, options.Selector.String(), finalErr)
+		} else {
+			klog.V(2).Infof("%s: %s: completed successfully", options.CallerName, options.Selector.String())
+		}
+	}()
+
+	// The periodic status dumps below only make sense when there is a
+	// single WaitForPods running, for example in the WaitForRunningPods
+	// measurements. When WaitForControlledPodsRunning invokes it for every
+	// single deployment in parallel, then the log output is too verbose.
+	klogVProgress := klog.V(2)
```

Comment:
While I agree that dumping all the data every 5s is way to verbose, I have used those logs in the past for debugging of some test cases.

Can we maybe simply change them to log only if they have changed  since the previous dump or something like that? Wouldn't that be sufficient?

Reply:
At -v2 it now only logs on changes. At -v5 you still get periodic dumps.

Reply:
Usually we are debugging already finished runs so there is no way to change verbosity flag for them.

Reply:
`SilentProgress: true` can also become `SilentProgress: false`, i.e. dump progress by default. Then when I run it, I can disable the excessive output. It just seemed to me that the other way around would be a more sensible default.